### PR TITLE
Thomas Casady Mid Mod

### PR DESCRIPTION
**Pull request by Tscasady**
I felt confident throughout the project, after finishing the extension work I added test cases for movies without actors and studios without movies.

### CODE REVIEW

File: app/controllers/movie_actors_controller.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class MovieActorsController < ApplicationController
+  def create
+    movie = Movie.find(params[:id])
+    actor = Actor.find(params[:actor_id])
+    MovieActor.create!(movie: movie, actor: actor)
```
Comment (cjsim89) [inline app/controllers/movie_actors_controller.rb:5]: Nice job using a 3rd controller! I always think this is the best course of action.

File: app/models/movie.rb [hunk 2]
```diff
@@ -1,3 +1,17 @@
...
+
+  def actors_by_age
+    self.actors.order(:age)
+  end
+
+  def average_actor_age
+    if self.actors.average(:age)
+      self.actors.average(:age)
+    else
+      0
```
Comment (cjsim89) [inline app/models/movie.rb:14]: This is interesting, I believe `.average` as an AR record will return 0 if there are no `actors` to run against. I might be wrong though - did you notice that it did return nil or something else, which made this necessary?

File: app/models/studio.rb [hunk 3]
```diff
@@ -1,3 +1,8 @@
 class Studio < ApplicationRecord
   has_many :movies
+  has_many :actors, through: :movies 
+
+  def uniq_actors
+    self.actors.uniq
```
Comment (cjsim89) [inline app/models/studio.rb:6]: `.uniq` in this context is actually a Ruby method; you are correct that we want unique values, but instead we want to use an AR method, which is `.distinct`. To see the difference, run this in `pry` and see what you get when you add `.to_sql`. It should error out. 
Then, try to run `self.actors.to_sql` and see what the difference is. I think doing that will illustrate why we want to use `.distinct` instead of `.uniq`.

File: app/views/movies/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+<%= "Movie title: #{@movie.title}" %><br>
+<%= "Created: #{@movie.creation_year}" %><br>
+<%= "Genre: #{@movie.genre}" %><br>
+
+<h2>Actors</h2>
+<% @movie.actors_by_age.each do |actor| %>
+  <%= actor.name %>
+<% end %>
+<p>Average Age of Actors: <%= @movie.average_actor_age.round(2) %></p>
```
Comment (cjsim89) [inline app/views/movies/show.html.erb:11]: I think we could probably have a debate/discussion about this, but in general a `.round` method I think should belong in the model. It's not an AR method, but it is data logic in that we're changing the value. It is a slight difference, and I could also argue for this since we might want to show different values in different places in the application, but for now try to put your data logic in the model method.

File: app/views/movies/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+<%= form_with url: "/movie_actors/#{@movie.id}", method: :post do |form| %>
```
Comment (cjsim89) [inline app/views/movies/show.html.erb:13]: Nice job with this RESTful request!

File: config/routes.rb [hunk 6]
```diff
@@ -1,3 +1,8 @@
 Rails.application.routes.draw do
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  get '/studios', to: "studios#index"
+  get '/studios/:id', to: "studios#show"
+
+  get '/movies/:id', to: "movies#show"
+  post '/movie_actors/:id', to: "movie_actors#create"
```
Comment (cjsim89) [inline config/routes.rb:7]: Great job, these are all RESTful routes.

File: spec/features/movie/show_spec.rb [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    let!(:studio_2) { Studio.create!(name: "Film Productions", location: "Denver") }
+    let!(:movie_1) { studio_1.movies.create!(title: "Everything Everywhere", creation_year: 2022, genre: "Drama") }
+    let!(:movie_2) { studio_2.movies.create!(title: "Spiderman #189", creation_year: 2022, genre: "Tired") }
+    let!(:movie_3) { studio_1.movies.create!(title: "Die Hard", creation_year: 1988, genre: "Action") }
+    let!(:movie_4) { studio_2.movies.create!(title: "Moana", creation_year: 2016, genre: "Animated") }
+    let!(:movie_5) { studio_2.movies.create!(title: "Empty", creation_year: 2016, genre: "Empty") }
+    let!(:actor_1) { movie_1.actors.create!(name: "Michelle Yeoh", age: 60) }
+    let!(:actor_2) { movie_1.actors.create!(name: "Stephanie Hsu", age: 32) }
+    let!(:actor_3) { movie_1.actors.create!(name: "Ke Huy Quan", age: 51) }
+    let!(:actor_4) { movie_4.actors.create!(name: "Auli'i Cravalho", age: 22) }
```
Comment (cjsim89) [inline spec/features/movie/show_spec.rb:15]: Really good object setup in your tests; I love having multiple objects of each type to create new and interesting scenarios from.

File: spec/features/movie/show_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+  describe 'when a user visits a movie show page' do
+    it 'displays the movies title, creation year, and genre' do
+      visit "/movies/#{movie_1.id}"
+
+      expect(page).to have_content "Everything Everywhere"
+      expect(page).to have_content "Drama"
+      expect(page).to have_content "2022"
```
Comment (cjsim89) [inline spec/features/movie/show_spec.rb:23]: I'm not really sure from this test alone what the "Drama" and "2022" values represent - testing for 
```
expect(page).to have_content "Genre: Drama" 
expect(page).to have_content "Year: 2022" 
```
is much more descriptive. What I like to think is that (even if it's not true) a totally different developer isg oing to come in and write the views from the way I wrote my feature tests - I'd better be darn sure that they are descriptive enough to where they don't have to ping me on Slack and ask me wtf I'm trying to describe here :) If I've succeeded, then I (or whoever does the views) will have a very easy time writing the actual View code if my feature tests are super descriptive, accurate, and precise.

File: spec/features/movie/show_spec.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    it 'displays the average age of all the movies actors' do
+      visit "/movies/#{movie_1.id}"
+
+      expect(page).to have_content "Average Age of Actors: 47.67"
+    end
+
+    it 'has a form to add an actor to the movie' do
+      visit "/movies/#{movie_1.id}"
+
+      expect(page.has_field?(:actor_id)).to be true
```
Comment (cjsim89) [inline spec/features/movie/show_spec.rb:43]: This is fine, I think you can also use `expect(page).to have_field(:actor_id)`.

File: spec/features/movie/show_spec.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    it 'can display an updated list of actors when the form is filled in with an actor id' do
+      visit "/movies/#{movie_4.id}"
+
+      expect(page).to_not have_content "Michelle Yeoh"
+
+      fill_in :actor_id, with: actor_1.id
+      click_button "Add Actor" 
+
+      expect(current_path).to eq "/movies/#{movie_4.id}"
+      expect(page).to have_content "Michelle Yeoh"
```
Comment (cjsim89) [inline spec/features/movie/show_spec.rb:55]: Good job - you're confirming the negative case so that the positive case on line 55 carries weight. 👍

File: spec/features/studio/index_spec.rb [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+    it 'displays all the studios movies with their title, creation year, and genre' do
+      visit '/studios'
+
+      within("#studio_#{studio_1.id}") do
+        expect(page).to have_content "Everything Everywhere" 
+        expect(page).to have_content "Die Hard" 
+        expect(page).to have_content "2022" 
+        expect(page).to have_content "1988" 
+        expect(page).to have_content "Action" 
+        expect(page).to have_content "Drama"
```
Comment (cjsim89) [inline spec/features/studio/index_spec.rb:35]: Good job using `within` blocks here; you can strengthen them even more by testing for some negative cases where applicable - for example, within studio_1's section, we would not expect to see `Spiderman` here. So, we should confirm that by writing something like:
```
expect(page).to_not have_content("Spiderman")
```

File: spec/features/studio/show_spec.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  
+    it 'displays a unique list of all of the actors that have worked for the studio' do
+      visit "/studios/#{studio_1.id}"
+
+      expect(page).to have_content "Bruce Willis" 
+      expect(page).to have_content "Bonnie Bedlia" 
+      expect(page).to have_content "Michelle Yeoh" 
+      expect(page).to have_content "Stephanie Hsu" 
+      expect(page).to have_content "Ke Huy Quan" 
+      expect(page).to_not have_content "Auli'i Cravalho"
```
Comment (cjsim89) [inline spec/features/studio/show_spec.rb:37]: For a list to be confirmed unique, we'd also want to make sure that each string only appears once. For that, you can add on `count: 1` to the end of each line. 
Ex: 
```
expect(page).to have_content("Bruce Willis"), count: 1
```

File: spec/models/movie_spec.rb [hunk 13]
```diff
@@ -1,7 +1,31 @@
...
+    it 'returns the movies actors from youngest to oldest' do
+      expect(movie_1.actors_by_age).to eq([actor_2, actor_3, actor_1])
+      expect(movie_2.actors_by_age).to eq([])
+    end
+  end
+
+  describe '#average_actor_age' do
+    it 'returns the movies average actor age' do
+      expect(movie_1.average_actor_age.round(2)).to eq(47.67)
+      expect(movie_2.average_actor_age.round(2)).to eq(0)
```
Comment (cjsim89) [inline spec/models/movie_spec.rb:28]: Cool - keep this test and remove the `if/else` from the model, I'm genuinely curious if that is necessary

File: spec/models/studio_spec.rb [hunk 14]
```diff
@@ -3,5 +3,31 @@
...
+    let!(:actor_4) { movie_4.actors.create!(name: "Auli'i Cravalho", age: 22) }
+    let!(:actor_5) { movie_3.actors.create!(name: "Bruce Willis", age: 67) }
+    let!(:actor_6) { movie_3.actors.create!(name: "Bonnie Bedlia", age: 74) }
+
+    let!(:ma1) { MovieActor.create!(movie: movie_3, actor: actor_3) }
+    
+    it 'returns a list of unique actors that work for the studio' do
+      expect(studio_1.uniq_actors).to eq([actor_1, actor_2, actor_3, actor_5, actor_6])
+      expect(studio_2.uniq_actors).to eq([actor_4])
+      expect(studio_3.uniq_actors).to eq([])
```
Comment (cjsim89) [inline spec/models/studio_spec.rb:30]: Excellent example of a strong model test! I like testing for a value, a different value, and then an empty one. Keep this up!! 👍
